src/optimizer/statistics/stats_calculator.cpp

```diff
@@ -0,0 +1,272 @@
+                                        std::unordered_map<std::string, std::unique_ptr<ColumnStats>> *stats) {
+  TERRIER_ASSERT(col->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE, "Expected ColumnValue");
+  auto tv_expr = col.CastManagedPointerTo<parser::ColumnValueExpression>();
+  if (table_stats->GetColumnCount() == 0 || !table_stats->HasColumnStats(tv_expr->GetColumnOid())) {
+    // We do not have stats for the table yet, use default value
+    stats->insert(std::make_pair(tv_expr->GetFullName(), CreateDefaultStats(tv_expr)));
+  } else {
+    stats->insert(std::make_pair(tv_expr->GetFullName(),
+                                 std::make_unique<ColumnStats>(*table_stats->GetColumnStats(tv_expr->GetColumnOid()))));
+  }
+}
+
+size_t StatsCalculator::EstimateCardinalityForFilter(
+    size_t num_rows, const std::unordered_map<std::string, std::unique_ptr<ColumnStats>> &predicate_stats,
+    const std::vector<AnnotatedExpression> &predicates) {
+  // First, construct the table stats as the interface needed it to compute selectivity
+  // TODO(boweic): We may want to modify the interface of selectivity computation to not use table_stats
+  std::vector<common::ManagedPointer<ColumnStats>> predicate_stats_vec;
+  auto table_stats = new TableStats();
+  for (auto &predicate : predicate_stats) {
+    table_stats->AddColumnStats(std::make_unique<ColumnStats>(*predicate.second));
+  }
+
+  double selectivity = 1.F;
+  for (auto &annotated_expr : predicates) {
+    // Loop over conjunction exprs
+    selectivity *=
+        CalculateSelectivityForPredicate(common::ManagedPointer<TableStats>(table_stats), annotated_expr.GetExpr());
+  }
+
+  // Update selectivity
+  return static_cast<size_t>(static_cast<double>(num_rows) * selectivity);
+}
+
+// Calculate the selectivity given the predicate and the stats of columns in the
+// predicate
+double StatsCalculator::CalculateSelectivityForPredicate(common::ManagedPointer<TableStats> predicate_table_stats,
+                                                         common::ManagedPointer<parser::AbstractExpression> expr) {
+  double selectivity = 1.F;
+  if (predicate_table_stats->GetColumnCount() == 0) {
+    return selectivity;
+  }
+
+  // Base case : Column Op Val
+  if ((expr->GetChild(0)->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE &&
+       (expr->GetChild(1)->GetExpressionType() == parser::ExpressionType::VALUE_CONSTANT ||
+        expr->GetChild(1)->GetExpressionType() == parser::ExpressionType::VALUE_PARAMETER)) ||
+
+      (expr->GetChild(1)->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE &&
+       (expr->GetChild(0)->GetExpressionType() == parser::ExpressionType::VALUE_CONSTANT ||
+        expr->GetChild(0)->GetExpressionType() == parser::ExpressionType::VALUE_PARAMETER))) {
+    int right_index = expr->GetChild(0)->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE ? 1 : 0;
+    auto left_expr = expr->GetChild(1 - right_index);
```

Comment:
Forgot to mention... what's the meaning of left and right here? Please add some comments.

Reply:
Comments added.